srml/transaction-payment/src/lib.rs

```diff
@@ -0,0 +1,456 @@
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! # Transaction Payment Module
+//!
+//! This module provides the basic logic needed to pay the absolute minimum amount needed for a
+//! transaction to be included. This includes:
+//!   - _weight fee_: A fee proportional to amount of Weight a transaction consumes. (TODO: explain
+//!     what weight is or link somewhere)
+//!   - _length fee_: A fee proportional to the encoded length of the transaction.
+//!   - _tip_: An optional tip. Tip increases the priority of the transaction, giving it a higher
+//!     chance to be included by the transaction queue.
+//!
+//! Additionally, this module allows one to configure:
+//!   - The mapping between one unit of weight to one unit of fee via [`WeightToFee`].
+//!   - A means of updating the fee for the next block, via defining a multiplier, based on the
+//!     final state of the chain at the end of the previous block. This can be configured via
+//!     [`FeeMultiplierUpdate`]
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use rstd::prelude::*;
+use codec::{Encode, Decode};
+use support::{
+	decl_storage, decl_module,
+	traits::{Currency, Get, OnUnbalanced, ExistenceRequirement, WithdrawReason},
+};
+use sr_primitives::{
+	Fixed64,
+	transaction_validity::{
+		TransactionPriority, ValidTransaction, InvalidTransaction, TransactionValidityError,
+		TransactionValidity,
+	},
+	traits::{Zero, Saturating, SignedExtension, SaturatedConversion, Convert},
+	weights::{Weight, DispatchInfo},
+};
+
+type Multiplier = Fixed64;
+type BalanceOf<T> =
+	<<T as Trait>::Currency as Currency<<T as system::Trait>::AccountId>>::Balance;
+type NegativeImbalanceOf<T> =
+	<<T as Trait>::Currency as Currency<<T as system::Trait>::AccountId>>::NegativeImbalance;
+
+pub trait Trait: system::Trait {
+	/// The currency type of the chain.
+	type Currency: Currency<Self::AccountId>;
+
+	/// Handler for the unbalanced reduction when taking transaction fees.
+	type OnTransactionPayment: OnUnbalanced<NegativeImbalanceOf<Self>>;
+
+	/// The fee to be paid for making a transaction; the base.
+	type TransactionBaseFee: Get<BalanceOf<Self>>;
+
+	/// The fee to be paid for making a transaction; the per-byte portion.
+	type TransactionByteFee: Get<BalanceOf<Self>>;
+
+	/// Convert a weight value into a deductible fee based on the currency type.
+	type WeightToFee: Convert<Weight, BalanceOf<Self>>;
+
+	/// Update the multiplier of the next block, based on the previous block's weight.
+	// TODO: maybe this does not need previous weight and can just read it
```

Comment:
But for polkadot we do need previous block weight isn't it ?
then we can remove this todo, someone can implement a constant function indeed and this is fine.

Reply:
I am still not sure about it. 

If possible, I would like to remove the hardcoded function api and make it just a simple `Convert<Fixed64, Fixed64>`. But first, a technical issue that I haven't really looked into: storage is not readable outside of externalities and I haven't looked into how it can be done here.

Reply:
oh ok that make sense then,

also this convert is called inside externalities.

Reply:
This convert is called inside externalities, but if I want to read directly from the storage of `System` in `node/runtime/src/impl.rs`, then I am not in externalities (right?) and for this, unfortunately, we have to hardcode the weight as the argument. Hopefully this can be further refactored.

Reply:
you can do, if AllExtrinsicsWeight is public:
```rust
system::AllExtrinsicsWeight::get()
```
At runtime if it called inside some externalities then this is will run ok otherwise it will crash with an error.